### DISCUSSION TITLE
How to enrich traces with data from scoped service?

### DISCUSSION THREAD

**Discussion by BogdanGirnyk** [open-telemetry/opentelemetry-dotnet#3851] [Q&A]
Hi, I want to enrich traces with data related to current user.
I've tried Enriching Processor as described [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/README.md#enriching-processor).
The issue is that the `Processors` are registered in DI as a singletons so there is no suitable way to inject user data that is scoped.
Could you advice the correct way to implement this task?

**Comment by cijothomas** [comment]
Have you looked at using the Enrich option provided by the instrumentation libraries?

**Reply by BogdanGirnyk** [reply]
All instrumentation Readme's state that:
_[Processor](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/README.md#processor), is the general extensibility point to add additional properties to any activity. The Enrich option is specific to instrumentation._
Beside of that, there is no way to inject custom service to instrumentation libraries enrichers. We can operate only with library-specific objects(e. g. SqlCommand for OpenTelemetry.Instrumentation.SqlClient)
